### PR TITLE
Replace `TmpDirCreator` by `TmpDirProvider `

### PR DESCRIPTION
**Pull request by theofidry**
With #869 merged, `TmpDirCreator` is now only used for our "production" code, i.e. is no longer used for the tests.

As a result, I'm taking this opportunity to change it: indeed I think the class (and its testing) is overly complex because besides providing a temporary path, it attempts to create it. This also has the unpleasant side effect of potentially creating a temporary directory whenever a `ConfigurationFactory` is used to create a configuration.

In other words: this PR simplifies `TmpDirCreator` by making it responsible only to provide a temporary directory, making the creation of `Configuration` via `ConfigurationFactory` side-effect free.

### CODE REVIEW

File: src/Configuration/ConfigurationFactory.php [hunk 1]
```diff
@@ -103,7 +103,7 @@ public function create(
             // calculation" and the "tmp dir create" in order to not have
```
Comment (maks-rafalko) [inline src/Configuration/ConfigurationFactory.php]: is this comment still needed?

File: src/Command/InfectionCommand.php [hunk 2]
```diff
@@ -285,6 +286,11 @@ private function startUp(): TestFrameworkAdapter
 
         $this->includeUserBootstrap($config);
 
+        /** @var Filesystem $fileSystem */
+        $fileSystem = $this->container['filesystem'];
+
+        $fileSystem->mkdir($config->getTmpDir());
```
Comment (maks-rafalko) [inline src/Command/InfectionCommand.php]: do we need `, 0777` as a second parameter (as it was in the `TmpDirCreator`)?

Comment (theofidry) [reply src/Command/InfectionCommand.php]: we can put it but it's the default so not a necessity
